File: ADF1/31-05-2021/aptechexcersise/src/main/java/com/mycompany/aptechexcersise/models/Shape.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.mycompany.aptechexcersise.models;

public abstract class Shape {
    private String name;
    private String color;
    //1 parent/super co nhieu sub-class
    //1 lop con co the co 2 lop cha ? NO(java) !,
    //ham khoi tao
    public Shape(String name, String color) {
        this.name = name;
        this.color = color;
    }

    protected String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }
    public void doXX(){
        System.out.println("Do xx(in Shape)");
    }
    public abstract Double calculateArea();//ko can thuc thi
    //multi-level = class A cha cua B, C la con cua B =>ok

    @Override
    public String toString() {
        return "name='" + name + '\'' +
                ", color='" + color + '\'';
    }
}

File: PHP/24-02-2021/Shape.php
<?php
class Shape {
    //Shape la lop cha(base class, super class) cua Rectangle
    private $name;
    private $color;
    
    //getter
    protected function get_name() {
        //phuong thuc/thuoc tinh protected co the goi duojc tu lop con
        //Ko goi duoc phuong thuc/thuoc tinh protected o ngoai class
        return $this->name;
    }
    public function get_color() {
        return $this->color;
    }
    //setter
    public function set_name($name) {
        $this->name = $name;
    }
    public function set_color($color) {
        $this->color = $color;
    }
}
?>